nixos/modules/services/network-filesystems/iroh.nix

```diff
@@ -0,0 +1,275 @@
+      # TODO use regex to match URL, including https:// prefix!
+      url = mkOption {
+        description = "The URL of the relay node.";
+        type = types.str;
+      };
+
+      stunOnly = mkOption {
+        description = "Only use STUN for traversal of NAT gateways.";
+        type = types.bool;
+      };
+
+      stunPort = mkOption {
+        description = "Port at which STUN service is listening.";
+        type = types.port;
+      };
+    };
+  };
+
+  # The total set of relays is the merge of relayNodes (which are the default nodes provided by
+  # iroh, if not replaced) and any extra nodes defined by the user.
+  relaySettings = cfg.relayNodes // cfg.extraRelayNodes;
+
+  # The config file should be located in the data directory and is NOT created by default. All
+  # values are optional, including the file itself.
+  configFormat = pkgs.formats.toml { };
+  #configFile = configFormat.generate "${cfg.dataDir}/iroh.config.toml" relaySettings;
+  configFile = configFormat.generate "/var/lib/iroh/iroh.config.toml" relaySettings;
+in
+{
+  # A NixOS module is a function that takes an environment and returns an attrset with the imports,
+  # options, and config fields. The provided parameters include:
+  # - config: an attrset with all the values of the configuration options
+  # - pkgs: all the packages defined in nixpkgs
+  # - lib: the utility library that comes with nixpkgs; it's commonly required so it is imported
+  #   using 'with lib'
+  # - other fields not usually needed by user modules
+
+  # A list of other modules to import. Only necessary if the imported module isn't already in
+  # module-list.nix, e.g. when writing several related modules, where only one of them is added to
+  # the global module list. When using "imports' the arguments to the current module (config, lib,
+  # pkgs) will also be passed to the submodules.
+  imports = [
+    #./other-module.nix
+  ];
+
+  # Option declarations for our module.
+  options = {
+    services.iroh = {
+
+      # By convention, modules have an "enable" option that determines whether they are active or
+      # not.
+      enable = mkEnableOption "Iroh is a protocol for syncing bytes. Bytes of any size, across any number of devices. Use iroh to create apps that scale different.";
+
+      package = mkPackageOption pkgs "iroh" { };
+
+      # NB: This is READ ONLY because there's no way to configure the listen port; iroh exposes as
+      # few configuration knobs as possible. The default port can be read from this read-only
+      # option, but if a different port is chosen by the service it has to be retrieved another way (?).
+      listenPort = mkOption {
+        description = "Address iroh listens on for connections from other nodes. If the port is taken iroh will choose a random port to listen on, so it can't be configured. This port is NOT automatically opened in the firewall.";
+        type = types.port;
+        readOnly = true;
+        default = listenPort;
```

Comment:
this value may not be correct so I'm not sure there is much value in having this option